gymnasium/envs/tabular/blackjack.py

```diff
@@ -0,0 +1,477 @@
+    the value of the dealer's one showing card (1-10 where 1 is ace),
+    and whether the player holds a usable ace (0 or 1).
+
+    This environment corresponds to the version of the blackjack problem
+    described in Example 5.1 in Reinforcement Learning: An Introduction
+    by Sutton and Barto (http://incompleteideas.net/book/the-book-2nd.html).
+
+    ### Rewards
+    - win game: +1
+    - lose game: -1
+    - draw game: 0
+    - win game with natural blackjack:
+
+        +1.5 (if <a href="#nat">natural</a> is True)
+
+        +1 (if <a href="#nat">natural</a> is False)
+
+    ### Arguments
+
+    ```
+    gym.make('Jax-Blackjack-v0', natural=False, sutton_and_barto=False)
+    ```
+
+    <a id="nat">`natural=False`</a>: Whether to give an additional reward for
+    starting with a natural blackjack, i.e. starting with an ace and ten (sum is 21).
+
+    <a id="sutton_and_barto">`sutton_and_barto=False`</a>: Whether to follow the exact rules outlined in the book by
+    Sutton and Barto. If `sutton_and_barto` is `True`, the keyword argument `natural` will be ignored.
+    If the player achieves a natural blackjack and the dealer does not, the player
+    will win (i.e. get a reward of +1). The reverse rule does not apply.
+    If both the player and the dealer get a natural, it will be a draw (i.e. reward 0).
+
+    ### Version History
+    * v0: Initial version release (0.0.0), adapted from original gym blackjack v1
+    """
+
+    action_space = spaces.Discrete(2)
+
+    observation_space = spaces.Box(
+        low=np.array([1, 1, 0]), high=np.array([32, 11, 1]), shape=(3,), dtype=np.int32
+    )
+
+    metadata = {
+        "render_modes": ["rgb_array"],
+        "render_fps": 4,
+    }
+
+    def __init__(self, natural: bool = False, sutton_and_barto: bool = True):
+        """Initializes Blackjack functional env."""
+        # Flag to payout 1.5 on a "natural" blackjack win, like casino rules
+        # Ref: http://www.bicyclecards.com/how-to-play/blackjack/
+        self.natural = natural
+        # Flag for full agreement with the (Sutton and Barto, 2018) definition. Overrides self.natural
+        self.sutton_and_barto = sutton_and_barto
+
+    def transition(
+        self, state: jnp.ndarray, action: Union[int, jnp.ndarray], key: PRNGKey
+    ):
+        """The blackjack environment's state transition function."""
+        env_state = jax.lax.cond(action, take, notake, (state, key))
+
+        hand_state, key = env_state
+        dealer_hand = hand_state[0]
+        player_hand = hand_state[1]
+        dealer_cards = hand_state[2]
+        player_cards = hand_state[3]
+
+        # note that only a bust or player action ends the round, the player
+        # can still request another card with 21 cards
+        done = (is_bust(player_hand) * action) + ((jnp.logical_not(action)) * 1)
+
+        new_state = (dealer_hand, player_hand, dealer_cards, player_cards, done)
```

Comment:
Could we use a `NamedTuple` as this is make code later easier to read with `state.dealer_hand` rather than `state[0]`

We might need to use a jax specify thing to do this

Reply:
changed to named tuple, Jax actually knows how to flatten namedtuples I think without requiring any custom stuff(we will find out when profiling the jitted code maybe).